### PR TITLE
Button to display Body in LightTheme

### PR DESCRIPTION
**Pull request by LunarX**

### CODE REVIEW

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadAdapter.kt [hunk 1]
```diff
@@ -79,39 +79,76 @@ class ThreadAdapter : RecyclerView.Adapter<ThreadViewHolder>(), RealmChangesBind
...
 
-    private fun ItemMessageBinding.loadBodyInWebView(body: Body?) {
+    private fun ThreadViewHolder.loadBodyInWebView(message: Message) = with(binding) {
         if (WebViewFeature.isFeatureSupported(WebViewFeature.ALGORITHMIC_DARKENING)) {
-            WebSettingsCompat.setAlgorithmicDarkeningAllowed(messageBody.settings, true)
+            WebSettingsCompat.setAlgorithmicDarkeningAllowed(messageBody.settings, isThemeTheSameMap[message.uid]!!)
         }
-        // TODO: Make prettier webview, Add button to hide / display the conversation inside message body like webapp ?
-        body?.let {
+        // TODO: Make prettier WebView, Add button to hide / display the conversation inside message body like webapp ?
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadAdapter.kt:145]:
```suggestion
        // TODO: Make prettier WebView, add button to hide/display the conversation inside Message body like WebApp ?
```

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadFragment.kt [hunk 2]
```diff
@@ -96,13 +96,18 @@ class ThreadFragment : Fragment() {
...
+            threadAdapter.isThemeTheSameMap = result.third
             observeMessagesLive()
             observeContacts()
             observeQuickActionBarClicks()
             observeOpenAttachment()
         }
 
+        mainViewModel.toggleLightThemeForMessage.observe(viewLifecycleOwner) {
+            threadAdapter.toggleLightMode(it)
+        }
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadFragment.kt:109]:
```suggestion
        mainViewModel.toggleLightThemeForMessage.observe(viewLifecycleOwner, threadAdapter::toggleLightMode)
```

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/MessageActionsBottomSheetDialog.kt [hunk 3]
```diff
@@ -37,6 +39,9 @@ class MessageActionsBottomSheetDialog : MailActionsBottomSheetDialog() {
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) = with(navigationArgs) {
         super.onViewCreated(view, savedInstanceState)
 
+        binding.lightTheme.isVisible = requireContext().isNightModeEnabled()
+        binding.lightTheme.setText(if (isThemeTheSame) R.string.actionViewInLight else R.string.actionViewInDark)
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/MessageActionsBottomSheetDialog.kt:43]:
```suggestion
        binding.lightTheme.apply {
            isVisible = requireContext().isNightModeEnabled()
            setText(if (isThemeTheSame) R.string.actionViewInLight else R.string.actionViewInDark)
        }
```

File: app/src/main/java/com/infomaniak/mail/utils/Extensions.kt [hunk 4]
```diff
@@ -163,14 +163,20 @@ fun Context.formatSubject(subject: String?): String {
...
+        head().appendElement("style").attr("type", "text/css").appendText(css).also { element ->
+            id?.let { element.id(it) }
+        }
         html()
     }
 }
 
+fun Context.readRawResource(cssResId: Int): String {
+    return Scanner(resources.openRawResource(cssResId)).useDelimiter("\\A").next()
+}
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/utils/Extensions.kt:178]:
```suggestion
fun Context.readRawResource(cssResId: Int): String = Scanner(resources.openRawResource(cssResId)).useDelimiter("\\A").next()
```

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadAdapter.kt [hunk 5]
```diff
@@ -79,39 +79,76 @@ class ThreadAdapter : RecyclerView.Adapter<ThreadViewHolder>(), RealmChangesBind
...
-        holder.bindAttachment(message)
-        loadBodyInWebView(message.body)
+        if (isExpandedMap[message.uid] == null) {
+            isExpandedMap[message.uid] = message.shouldBeExpanded(position, messages.lastIndex)
+        }
 
-        displayExpandedCollapsedMessage(message)
+        if (isThemeTheSameMap[message.uid] == null) {
+            isThemeTheSameMap[message.uid] = true
+        }
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadAdapter.kt:132]: This if can be one-lined

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadFragment.kt [hunk 6]
```diff
@@ -96,13 +96,18 @@ class ThreadFragment : Fragment() {
             val threadUid = result.first.uid
             setupUi(threadUid)
             setupAdapter(threadUid)
-            threadAdapter.expandedMap = result.second
+            threadAdapter.isExpandedMap = result.second
+            threadAdapter.isThemeTheSameMap = result.third
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadFragment.kt:100]: What about creating a data class for the results, to be a bit more specific than a triple ?

**Review by FabianDevel — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadFragment.kt (hunk 6 above)
Comment (LunarX) [reply app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadFragment.kt:100]: It's only used once here and all three values doesn't really represent a very well defined object

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]
Doesn't work on API 31.

**Review by KevinBoulongne — APPROVED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/MessageActionsBottomSheetDialog.kt [hunk 7]
```diff
@@ -42,6 +44,12 @@ class MessageActionsBottomSheetDialog : MailActionsBottomSheetDialog() {
             setMarkAsReadUi(isSeen)
             setFavoriteUi(isFavorite)
 
+            binding.lightTheme.apply {
+                isVisible = requireContext().isNightModeEnabled()
+                setText(if (isThemeTheSame) R.string.actionViewInLight else R.string.actionViewInDark)
+                setClosingOnClickListener { mainViewModel.toggleLightThemeForMessage.value = message }
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/MessageActionsBottomSheetDialog.kt:50]:
```suggestion
                if (requireContext().isNightModeEnabled()) {
                    isVisible = true
                    setText(if (isThemeTheSame) R.string.actionViewInLight else R.string.actionViewInDark)
                    setClosingOnClickListener { mainViewModel.toggleLightThemeForMessage.value = message }
                }
```

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

**Pull request merged**